flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java

```diff
@@ -0,0 +1,263 @@
+    private final ExecutorService asyncOperationsThreadPool;
+
+    /** scheduled executor, periodically trigger materialization. */
+    private final ScheduledExecutorService periodicExecutor;
+
+    private final CheckpointStreamFactory streamFactory;
+
+    private final Snapshotable keyedStateBackend;
+
+    private final StateChangelogWriter<ChangelogStateHandle> stateChangelogWriter;
+
+    private final AsyncExceptionHandler asyncExceptionHandler;
+
+    private final int allowedNumberOfFailures;
+
+    /** Materialization failure retries. */
+    private final AtomicInteger retries;
+
+    /** Making sure only one materialization on going at a time. */
+    private final AtomicBoolean materializationOnGoing;
+
+    private long materializedId;
+
+    private MaterializedState materializedState;
+
+    public PeriodicMaterializer(
+            MailboxExecutor mailboxExecutor,
+            ExecutorService asyncOperationsThreadPool,
+            Snapshotable keyedStateBackend,
+            CheckpointStorageWorkerView checkpointStorageWorkerView,
+            StateChangelogWriter<ChangelogStateHandle> stateChangelogWriter,
+            AsyncExceptionHandler asyncExceptionHandler,
+            MaterializedState materializedState,
+            long periodicMaterializeInitDelay,
+            long periodicMaterializeInterval,
+            int allowedNumberOfFailures,
+            boolean materializationEnabled) {
+        this.mailboxExecutor = mailboxExecutor;
+        this.asyncOperationsThreadPool = asyncOperationsThreadPool;
+        this.keyedStateBackend = keyedStateBackend;
+        this.stateChangelogWriter = stateChangelogWriter;
+        this.asyncExceptionHandler = asyncExceptionHandler;
+        this.periodicExecutor =
+                Executors.newSingleThreadScheduledExecutor(
+                        new ExecutorThreadFactory("periodic-materialization"));
+        this.streamFactory = shared -> checkpointStorageWorkerView.createTaskOwnedStateStream();
+        this.allowedNumberOfFailures = allowedNumberOfFailures;
+        this.materializationOnGoing = new AtomicBoolean(false);
+        this.retries = new AtomicInteger(allowedNumberOfFailures);
+
+        this.materializedId = 0;
+        this.materializedState = materializedState;
+
+        if (materializationEnabled) {
+            this.periodicExecutor.scheduleAtFixedRate(
+                    this::triggerMaterialization,
+                    periodicMaterializeInitDelay,
+                    periodicMaterializeInterval,
+                    TimeUnit.MILLISECONDS);
+        }
+    }
+
+    @VisibleForTesting
+    public void triggerMaterialization() {
+        mailboxExecutor.execute(
+                () -> {
+                    // Only one materialization ongoing at a time
+                    if (!materializationOnGoing.compareAndSet(false, true)) {
+                        return;
+                    }
```

Comment:
That's a good point!

Reply:
Triggering immediately seems tricky to implement because we need to wait for the async phase to finish. 
So I'd go with min-pause solution and sceduling the next materialization upon finishing the current one. Let's discuss it offline.